chainermn/communicators/mpi_communicator_base.py

```diff
@@ -617,23 +618,77 @@ def scatter(self, xs, root=0):
                 root)
             return rbuf.reshape(shape)
 
+    def _check_obj_type_for_chainerx(self, obj):
+        # Do NOT support chainerx ndarray with CUDA
+        # for the following reason:
+        # (1) mpi4py.send pickles the object
+        # (2) chainerx.ndarray preserves CUDA
+        # device internally when pickled
+        # (3) An error will occur when an ndarray is unpickled in another
+        #     process
+        #
+        if None is obj:
+            return False
+
+        # check collections of list, tuple and set
+        elif type(obj) in [list, tuple, set]:
+            for item in obj:
+                xp = chainer.backend.get_array_module(item)
+                # DO NOT use device.backend.name as
+                # 'ChainerxDevice' object has no attribute 'backend'
+                if xp == chainerx and item.device.name.startswith('cuda'):
+                    return True
+
+        # check dict
+        elif type(obj) is dict:
+            for key, value in obj.items():
+                xp = chainer.backend.get_array_module(key)
+                if xp == chainerx and key.device.name.startswith('cuda'):
+                    return True
+
+                xp = chainer.backend.get_array_module(value)
+                if xp == chainerx and value.device.name.startswith('cuda'):
+                    return True
+        else:
```

Comment:
How about checking if `obj` is array object of `numpy`, 'cupy', and `chainerx`?

Reply:
`chainer.backend.get_array_module` can accept any type and return `numpy` for types that are not cupy nor chainerx. So I don't think we have to check if they beforehand.